### ISSUE TITLE
【效果讨论帖】大家来分享下预训练或者SFT后的效果，互相对比一下吧？

### ISSUE THREAD

**Issue by OKC13** [jingyaogong/minimind-v#98]
我预训练后效果如下，感觉视觉和语言能力确实有一定对齐，但是语言能力实在太弱了导致的。预训练LOSS也不下降，大家也是吗：
Epoch:[4/4](3000/4652) loss:3.015322 lr:0.000005966081 epoch_Time:9.0min:
Epoch:[4/4](3100/4652) loss:3.075659 lr:0.000005853312 epoch_Time:9.0min:
Epoch:[4/4](3200/4652) loss:3.387815 lr:0.000005747425 epoch_Time:8.0min:
Epoch:[4/4](3300/4652) loss:2.637612 lr:0.000005648451 epoch_Time:8.0min:
Epoch:[4/4](3400/4652) loss:3.116996 lr:0.000005556418 epoch_Time:7.0min:
Epoch:[4/4](3500/4652) loss:3.561009 lr:0.000005471353 epoch_Time:7.0min:
Epoch:[4/4](3600/4652) loss:3.270156 lr:0.000005393278 epoch_Time:6.0min:
Epoch:[4/4](3700/4652) loss:3.423081 lr:0.000005322218 epoch_Time:6.0min:
Epoch:[4/4](3800/4652) loss:3.031712 lr:0.000005258191 epoch_Time:5.0min:
Epoch:[4/4](3900/4652) loss:3.041336 lr:0.000005201216 epoch_Time:4.0min:
Epoch:[4/4](4000/4652) loss:3.222608 lr:0.000005151310 epoch_Time:4.0min:
Epoch:[4/4](4100/4652) loss:3.332428 lr:0.000005108486 epoch_Time:3.0min:
Epoch:[4/4](4200/4652) loss:3.198358 lr:0.000005072757 epoch_Time:3.0min:
Epoch:[4/4](4300/4652) loss:3.342056 lr:0.000005044133 epoch_Time:2.0min:
Epoch:[4/4](4400/4652) loss:3.185369 lr:0.000005022623 epoch_Time:1.0min:
Epoch:[4/4](4500/4652) loss:3.420405 lr:0.000005008231 epoch_Time:1.0min:
Epoch:[4/4](4600/4652) loss:3.302365 lr:0.000005000963 epoch_Time:0.0min:
Epoch:[4/4](4651/4652) loss:2.936298 lr:0.000005000000 epoch_Time:0.0min:

<img width="481" height="672" alt="Image" src="https://github.com/user-attachments/assets/d1e27c9c-387b-4bf9-b2b9-30d14a8a5d00" />

**Comment by jingyaogong** [comment]
简单来说，你的训练没收敛，loss参考：

<img width="1522" height="1328" alt="Image" src="https://github.com/user-attachments/assets/0918e867-aee8-4e2d-adcc-b7117e58e03c" />

检查你48.78M的定制化LLM

**Comment by jingyaogong** [comment]
@chenkang455 

<img width="1682" height="926" alt="Image" src="https://github.com/user-attachments/assets/3d1b6455-bcd0-447c-b4ae-2a93ceab89d0" />

以上这里是有前提的，loss不能完全参考这里，可以自行在`trainer_utils.py`中打开更多可训练参数（也可以是q v o proj层）例如：

```
    # pretrain阶段：解冻最后1层
    last_layer_idx = vlm_config.num_hidden_layers - 1
    for name, param in model.model.named_parameters():
        if f'layers.{last_layer_idx}.' in name:
            param.requires_grad = True
```

注意看，现在的默认设置里，pretrain 梯度只打开了proj层，这个loss很合理，目的是不破坏任何LLM知识，走势是：

<img width="2174" height="630" alt="Image" src="https://github.com/user-attachments/assets/a3d50ff7-52c6-4af2-91e8-25370f15ba1f" />

测试是：

```
python eval_vlm.py --hidden_size 768 --num_hidden_layers 16 --weight pretrain_vlm
VLM模型参数: 104.62 M(illion)
[图像]: #U57ce#U5e02#U8f66#U6c34#U9a6c#U9f99-city-traffic.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 从高处望去的城市景色。



[图像]: #U592a#U7a7a#U5b87#U822a#U5458-Astronaut-Space.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 旅行者在天空中看着一架飞机,飞机在空中飞行



[图像]: #U5c0f#U72d7#U7f8e#U5973#U6d77#U8fb9-Dog-Woman-Sea.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人坐在沙发上,坐在海滩上,在沙滩上看着海浪。



[图像]: #U5f69#U8679#U7011#U5e03-Rainbow-Falls.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 水是水的冰冻景象。



[图像]: #U6905#U5b50#U8001#U4eba#U770b#U4e66-Chair-Elderly-Reading.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人坐在书架上看书



[图像]: #U718a#U732b#U8349#U5730-Panda-Grassland.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 我是动物学家。



[图像]: #U81ea#U884c#U8f66#U9c9c#U82b1-Bicycle-Flowers.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 我爱在花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园里花园



[图像]: #U821e#U8e48-dance.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人在舞台上演出。
```


设置最后一层可训练后：

<img width="2182" height="628" alt="Image" src="https://github.com/user-attachments/assets/7edef0bf-925b-4b41-b092-5b2f67ad3ea0" />


```
python eval_vlm.py --hidden_size 768 --num_hidden_layers 16 --weight pretrain_vlm
VLM模型参数: 104.62 M(illion)
[图像]: #U57ce#U5e02#U8f66#U6c34#U9a6c#U9f99-city-traffic.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 从右边看,从左边看



[图像]: #U592a#U7a7a#U5b87#U822a#U5458-Astronaut-Space.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 宇航员的太空船



[图像]: #U5c0f#U72d7#U7f8e#U5973#U6d77#U8fb9-Dog-Woman-Sea.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人,在海滩上坐着,在沙滩上



[图像]: #U5f69#U8679#U7011#U5e03-Rainbow-Falls.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 白天在山脊上方的彩虹



[图像]: #U6905#U5b50#U8001#U4eba#U770b#U4e66-Chair-Elderly-Reading.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人坐在花园的长凳上,戴着一本书



[图像]: #U718a#U732b#U8349#U5730-Panda-Grassland.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 白色背景的可爱的小姑娘



[图像]: #U81ea#U884c#U8f66#U9c9c#U82b1-Bicycle-Flowers.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人在花园里散步



[图像]: #U821e#U8e48-dance.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 一个人在舞台上表演舞蹈。
```

**Comment by chenkang455** [comment]
@jingyaogong 好的好的十分感谢！！！谢谢您的帮助！困扰了好几天为什么loss下不去，原来是这样，非常感谢

**Comment by jingyaogong** [comment]
> [@jingyaogong](https://github.com/jingyaogong) 好的好的十分感谢！！！谢谢您的帮助！困扰了好几天为什么loss下不去，原来是这样，非常感谢

客气了😊 默认设置确实有歧义

基于此SFT 1 epoch后：
```
python eval_vlm.py --hidden_size 768 --num_hidden_layers 16 --weight sft_vlm
VLM模型参数: 104.62 M(illion)
[图像]: #U57ce#U5e02#U8f66#U6c34#U9a6c#U9f99-city-traffic.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 图片中，一辆红白相间的红色汽车停在城市街道上，营造出一种繁忙的气氛。在场景中可以看到几辆车，其中一辆靠近画面中心，另一辆则位于画面左侧。
除了汽车和行人，还有一些其他的车辆在场景中出现。有些车似乎停在路边，而其他的则散落在场景各处。这幅图像展示了城市的多彩景色，突出了繁忙的都市环境。



[图像]: #U592a#U7a7a#U5b87#U822a#U5458-Astronaut-Space.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 图片中，一只红色的大红巨型红外探测器正从一根金属球上飞出，它位于画面左下角附近。这只红色的巨型红外探测器看起来很聪明，可能是为了更容易地接近和接近它。
该图像描绘了一个有趣的场景，展示了红外探测器与球之间的互动，显示了它可能用于通信和导航。此外，还能看到一个装满各种物品的容器，如书籍、行李箱和书籍等，这表明这是一次有趣的互动活动。



[图像]: #U5c0f#U72d7#U7f8e#U5973#U6d77#U8fb9-Dog-Woman-Sea.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 图片中，一个女人站在海滩上，身边有几个人，其中一些人坐在沙滩上享受着阳光和沙滩。她似乎正在享受海滩时光，一边享受阳光和大海。



[图像]: #U5f69#U8679#U7011#U5e03-Rainbow-Falls.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 照片中，一大片平静的水面拍摄着一座桥，这座桥上有几个人，他们坐在上面，站着看着它。



[图像]: #U6905#U5b50#U8001#U4eba#U770b#U4e66-Chair-Elderly-Reading.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 照片中，一个女人坐在一张长椅上，背着背包，手里拿着一把椅子。她似乎在享受着这个小集市，周围是树木和一座桥。
在背景中，有一台电视机。背景中可以看到一辆卡车。总共有三个人，其中两个人靠近画面中心，另外两个人也出现在背景中。



[图像]: #U718a#U732b#U8349#U5730-Panda-Grassland.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 图片中，一只棕色的猫正坐在一块黑色的棕色小草上。它似乎正从黑色小草中走出来，似乎正在向外看。这只猫似乎正在享受它在草地上吃草的乐趣，看着它。



[图像]: #U81ea#U884c#U8f66#U9c9c#U82b1-Bicycle-Flowers.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 照片中，一个女人坐在一栋大楼的长凳上，手里拿着一把伞。她正坐在一块长凳上，周围是树木。
在背景中，可以看到一条长长的街道。一个女人站在她旁边的位置，可能正在观察附近的人们。



[图像]: #U821e#U8e48-dance.jpg
👶: 仔细看一下这张图：\n\n<image>\n\n描述一下这个图像的内容。
🤖️: 照片中，一个女人站在一面红色或粉红色的裙子前，手里拿着一把蓝色的大剪刀。她穿着一件蓝色连衣裙，手里拿着一件大剪刀。她穿着一件白色衬衫，可能是上衣或裤子。
在画面中，可以看到一只手拿着一个小女孩，可能是她的妈妈或孩子。她似乎正在等待一个陌生的男人到来，可能是在等待她的新朋友或者正在与他们互动。场景中还出现了一只鸟，这可能表示一个场景或一个有趣的事件。
```

测试了下FT效果上没大问题

---

https://github.com/jingyaogong/minimind-v/commit/3684285a914c90dcfa9edbf97cb02b105ef8ba7c

已更新parquet数据集格式，免大量小文件解压，且去掉了torch.cuda.empty_cache()频繁释放导致利用率低，以及dataloader速度等小问题

训练可能会加速很多，可更快的复现下以上问题是否解决，推荐尝试，祝好～
